### ISSUE TITLE
[BUG] window not darwn possibly due to size set to 0x0

### ISSUE THREAD

**Issue by VuiMuich** [elkowar/eww#281]
## Checklist before submitting an issue
- [x] I have searched through the existing [closed and open issues](https://github.com/elkowar/eww/issues?q=is%3Aissue) for eww and made sure this is not a duplicate
- [x] I have specifically verified that this bug is not a common [user error](https://github.com/elkowar/eww/issues?q=is%3Aissue+label%3Ano-actual-bug+is%3Aclosed)
- [x] I am providing as much relevant information as I am able to in this bug report (Minimal config to reproduce the issue for example, if applicable)

## Describe the bug
Earlier I reported, I am experiencing the zombie windows as described in #255 but in the meantime I believe this is actually a different issue.

The issue is, that windows are not drawn and when several windows should be open, it varies on each reload which window is actually drawn.

Snippet from `eww logs`:
```
Gtk-WARNING **: 17:25:31.006: Negative content width -25 (allocation 1, extents 13x13) while allocating gadget (node button, owner GtkButton)

(eww:4525): Gtk-WARNING **: 17:25:31.006: Negative content width -25 (allocation 1, extents 13x13) while allocating gadget (node button, owner GtkButton)

(eww:4525): Gtk-WARNING **: 17:25:31.006: Negative content width -25 (allocation 1, extents 13x13) while allocating gadget (node button, owner GtkButton)
```
```
(eww:36582): Gtk-WARNING **: 18:09:23.543: Negative content width -15 (allocation 5, extents 10x10) while allocating gadget (node box, owner GtkBox)

(eww:36582): Gtk-WARNING **: 18:09:23.543: Negative content height -15 (allocation 5, extents 10x10) while allocating gadget (node box, owner GtkBox)

(eww:36582): Gtk-WARNING **: 18:09:23.543: for_size smaller than min-size (1 < 3) while measuring gadget (node trough, owner GtkScale)

(eww:36582): Gtk-WARNING **: 18:09:25.545: Negative content width -15 (allocation 5, extents 10x10) while allocating gadget (node box, owner GtkBox)

(eww:36582): Gtk-WARNING **: 18:09:25.545: Negative content height -15 (allocation 5, extents 10x10) while allocating gadget (node box, owner GtkBox)

(eww:36582): Gtk-WARNING **: 18:09:25.545: for_size smaller than min-size (1 < 3) while measuring gadget (node trough, owner GtkScale)

(eww:36582): Gtk-WARNING **: 18:09:27.546: Negative content width -15 (allocation 5, extents 10x10) while allocating gadget (node box, owner GtkBox)

(eww:36582): Gtk-WARNING **: 18:09:27.546: Negative content height -15 (allocation 5, extents 10x10) while allocating gadget (node box, owner GtkBox)

(eww:36582): Gtk-WARNING **: 18:09:27.546: for_size smaller than min-size (1 < 3) while measuring gadget (node trough, owner GtkScale)

(eww:36582): Gtk-WARNING **: 18:09:29.547: Negative content width -15 (allocation 5, extents 10x10) while allocating gadget (node box, owner GtkBox)

(eww:36582): Gtk-WARNING **: 18:09:29.548: Negative content height -15 (allocation 5, extents 10x10) while allocating gadget (node box, owner GtkBox)

(eww:36582): Gtk-WARNING **: 18:09:29.548: for_size smaller than min-size (1 < 3) while measuring gadget (node trough, owner GtkScale)
```
```
$ eww windows
*tags
*left
notes
uhr
*right
```
For reference I removed my eww config and placed the example bar and also killed picom. This is when I saw this tiny sqare top center of my screen, and xprop on it revealed:
```
$ xprop
_NET_WM_DESKTOP(CARDINAL) = 0
_NET_WM_STATE(ATOM) = _NET_WM_STATE_ABOVE, _NET_WM_STATE_STICKY, _NET_WM_STATE_SKIP_TASKBAR, _NET_WM_STATE_SKIP_PAGER
WM_HINTS(WM_HINTS):
		Client accepts input or input focus: True
		Initial state is Normal State.
		window id # of group leader: 0x800001
_GTK_THEME_VARIANT(UTF8_STRING) = "dark"
_NET_WM_STRUT_PARTIAL(CARDINAL) = 0, 0, 57, 0, 0, 0, 0, 0, 0, 2559, 0, 0
_NET_WM_STRUT(CARDINAL) = 0, 0, 57, 0
XdndAware(ATOM) = BITMAP
_MOTIF_WM_HINTS(_MOTIF_WM_HINTS) = 0x2, 0x0, 0x0, 0x0, 0x0
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_DOCK
_NET_WM_SYNC_REQUEST_COUNTER(CARDINAL) = 8388613, 8388614
_NET_WM_USER_TIME_WINDOW(WINDOW): window id # 0x800004
WM_CLIENT_LEADER(WINDOW): window id # 0x800001
_NET_WM_PID(CARDINAL) = 36582
WM_LOCALE_NAME(STRING) = "de_DE.utf8"
WM_CLIENT_MACHINE(STRING) = "manjobook"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		program specified minimum size: 2304 by 43
		program specified base size: 0 by 0
		window gravity: Center
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS, _NET_WM_PING, _NET_WM_SYNC_REQUEST
WM_CLASS(STRING) = "eww-bar", "eww-bar"
WM_ICON_NAME(STRING) = "Eww - bar"
_NET_WM_ICON_NAME(UTF8_STRING) = "Eww - bar"
WM_NAME(STRING) = "Eww - bar"
_NET_WM_NAME(UTF8_STRING) = "Eww - bar"
```
## Reproducing the issue
I tried to `rustup update` and `cargo clean` before doing a fresh build of eww.
As mentioned above I tried my personal config as well as the example config.

## Expected behaviour
Windows should just be drawn.

## Additional context
I have not checked, if this appears in any other WM then leftwm yet (I am pretty positive though, that it is not leftwm as I reset my git repo tree to a version I new was working cleanly and had the same issue), also have not checked if this issue is know with gtk bug tracker.

**Comment by VuiMuich** [comment]
:thinking: tried bspwm and i3 and they seem to run ok.
so it has to be somewhat related to my leftwm session, but I don't see how atm.
Any further ideas how I could debug this?

**Comment by VuiMuich** [comment]
Ok, this is very weird.
When I set the windows to `wm-ignore = true` the are getting always drawn. but `reserve` does not work anymore, which kind of makes sense.
I have one window wich has `(deflisten workspace-layout 'leftwm-state -w 0 -s "{{workspace.layout}}"')` the return values in my WM config can be one of the following: `Monocle`, `MainAndDeck`, `LeftWiderRightStack`, `CenterMainBalanced`
When the window has size `0x0` and I change to one of the latter two layouts it suddenly gets drawn again.
The xprop output is unchanged by this.

Would it make sense to set the `program specified base size` in `WM_NORMAL_HINTS(WM_SIZE_HINTS)`?

**Comment by elkowar** [comment]
So is the issue that the window, when given size :width 0 :height 0 in the config, actually has the configured size of 0px*0px when initially opened, not respecting the fact that there is content to be rendered - and after changing the value the window size adjusts to fit the content?

**Comment by VuiMuich** [comment]
Windows actually have a configured size:
Here is a snippet of two of the windows:
```clojure
(defwindow left
           :monitor 0
           :windowtype "dock"
           :stacking "overlay"
           :geometry (geometry :x "21px"
                               :y "13px"
                               :width "295px"
                               ; :height "34px"
                               :height "44px" ; this is a workaround, as gtk is expanding the height to this anyways
                               :anchor "top left")
           :reserve (struts :side "top" :distance "55px")
    (sidestuff_left))

(defwindow tags
           :monitor 0
           :windowtype "dock"
           :geometry (geometry :x "0px"
                               :y "13px"
                               :width "19%"
                               ; :height "34px"
                               :height "44px" ; this is a workaround, as gtk is expanding the height to this anyways
                               :anchor "top center")
           :reserve (struts :side "top" :distance "55px")
           :wm-ignore true
    (tags))
```

**Comment by VuiMuich** [comment]
Could this be related to rust-nightly?
It seams odd to me, that rustup says I'm on `1..57.0-nightly (5ecc8ad84 2021-09-19)` but starshp prompt detects `1.56.0-nightly` for my repo direcotry..

**Comment by VuiMuich** [comment]
> ```
> WM_NORMAL_HINTS(WM_SIZE_HINTS):
> 		program specified minimum size: 320 by 44
> 		program specified base size: 0 by 0
> ```

Is this set somewhere in a depenency (gtk?) or are you setting these 'manually'? As I can't grep for it, probably dependency, right?

**Comment by ZetaNumbers** [comment]
Context: `LeftWM 0.2.8, Git-Hash: 3c80a4a`

@VuiMuich i changed the base size, it doesn't solve the issue.

Also i compared eww with polybar and lemonbar. They both have only a few lines of `xprop` output.

**Comment by ZetaNumbers** [comment]
Polybar and lemonbar use xcb to initialize the window i think.

**Comment by ZetaNumbers** [comment]
These are screenshots i took

`:wm-ignore false`:
![eww-respect-wm](https://user-images.githubusercontent.com/53300196/135093859-59630652-8aa7-49f7-a210-a6500d998e78.png)


`:wm-ignore true`:
![eww-ignore-wm](https://user-images.githubusercontent.com/53300196/135093662-b2cc92a0-dc9f-4be8-aada-9c37380f328e.png)

As you can see there's a small dot on the first screenshot. It is the same window which shows "example content".

**Comment by VuiMuich** [comment]
fixed with https://github.com/leftwm/leftwm/pull/498

**Comment by elkowar** [comment]
@VuiMuich do we still assume this to somehow also be a bug / non-spec-compliant behavior in Eww, or do you reckon this was just weird behavior from leftWM? I've kinda lost track of this issue over the past few weeks, I'm sorry ^^

**Comment by VuiMuich** [comment]
Well the fix on the `leftwm` side doesn't hurt, as it basically prohibits resizing for `dock` type windows.

Together with @ZetaNumbers and @AethanFoot (who then contributed leftwm/leftwm#498) we found that the i3 fix from #258 introduced this behaviour.
Specifically this `if` in `crates/eww/src/app.rs` lines 377ff:
```rust
    if actual_window_rect.x != gdk_origin.1 || actual_window_rect.y != gdk_origin.2 {
        gdk_window.move_(actual_window_rect.x, actual_window_rect.y);
    }
```
I am not sure at this point, if leftwm isn't operating on spec in this regard.
